### PR TITLE
backup: backup sub-ranges in one request

### LINKED ISSUE #13701: backup: backup sub-ranges in one request to support the checkpoint
## Development Task
When using checkpoint mode, BR divide a range into backed up and non backed up sub-ranges.
To avoid creating more request, the `BackupRequest` can hold the sub-ranges of the origin range in one transmission.

### PR DESCRIPTION
**Pull request by Leavrth**
Signed-off-by: Leavrth <jianjun.liao@outlook.com>

### What is changed and how it works?

Issue Number: Close #13701 

What's Changed: backup service now can back up multi ranges in one transmission.

```commit-message
```

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
- No code

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by Leavrth** [comment]
kvproto https://github.com/pingcap/kvproto/pull/996

File: components/backup/src/endpoint.rs [hunk 1]
```diff
@@ -1512,6 +1562,197 @@ pub mod tests {
...
+                    let start_key = if start_key.is_empty() {
+                        None
+                    } else {
+                        Some(Key::from_raw(start_key))
+                    };
+                    let end_key = if end_key.is_empty() {
+                        None
+                    } else {
+                        Some(Key::from_raw(end_key))
+                    };
```
Comment (YuJuncen) [inline components/backup/src/endpoint.rs:1590]: We can replace it with [`bool::then`](http://doc.rust-lang.org/1.65.0/core/primitive.bool.html#method.then)

File: components/backup/src/endpoint.rs [hunk 2]
```diff
@@ -996,17 +1028,34 @@ impl<E: Engine, R: RegionInfoProvider + Clone + 'static> Endpoint<E, R> {
...
+                ranges.push((start_key, end_key));
+            }
+            Arc::new(Mutex::new(Progress::new_with_ranges(
+                self.store_id,
+                ranges,
+                self.region_info.clone(),
+                codec,
+                request.cf,
+            )))
+        };
```
Comment (YuJuncen) [inline components/backup/src/endpoint.rs:1057]: Can we extract the huge if-else block into a function like `Self::progress_by_req(req: BackupRequest)`?

**Review by YuJuncen — APPROVED** [review]
rest LGTM

**Review by 3pointer — APPROVED** [review]

**Review by overvenus — APPROVED** [review]

**Pull request merged**